File: 57.Prim’sAlgo.py
# Prim’s Algorithm is a greedy algorithm used to find the Minimum Spanning Tree (MST) of a connected, 
# weighted, and undirected graph.
# It starts from an arbitrary node and repeatedly adds the smallest 
# edge that connects a visited vertex to an unvisited vertex.

# Explanation

# The goal of Prim’s Algorithm is to connect all vertices with the minimum total edge weight 
# without forming cycles.
# At each step:

# Start from any vertex.

# Pick the smallest edge connecting the current tree to a new vertex.

# Repeat until all vertices are included in the tree.

# It uses a priority queue (min-heap) to efficiently get the smallest edge at each step.

# Time Complexity: O(E log V)
# Space Complexity: O(V + E)

# Example Graph

# Consider this weighted undirected graph:

#      2
#  (A)---(B)
#   | \   |
#   3 \ \ 1
#   |   \ \
#  (C)---(D)
#      4

# Edges:
# A-B (2), A-C (3), A-D (1), B-D (2), C-D (4)

import heapq

def prims_algorithm(graph, start):
    visited = set()
    min_heap = [(0, start)]
    total_cost = 0

    while min_heap:
        weight, node = heapq.heappop(min_heap)

        if node in visited:
            continue

        visited.add(node)
        total_cost += weight

        for next_weight, neighbor in graph[node]:
            if neighbor not in visited:
                heapq.heappush(min_heap, (next_weight, neighbor))

    return total_cost

graph = {
    'A': [(2, 'B'), (3, 'C'), (1, 'D')],
    'B': [(2, 'A'), (2, 'D')],
    'C': [(3, 'A'), (4, 'D')],
    'D': [(1, 'A'), (2, 'B'), (4, 'C')]
}

mst_cost = prims_algorithm(graph, 'A')

print("Minimum Spanning Tree Cost:", mst_cost)
        

File: 57.Prim’sAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/* -----------------------------------------------------------------------  */

import java.util.*;

public class PrimsAlgorithm {

    // Edge class to represent graph edges
    static class Edge implements Comparable<Edge> {
        int vertex;
        int weight;

        Edge(int vertex, int weight) {
            this.vertex = vertex;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge other) {
            return this.weight - other.weight; // Sort by weight for PriorityQueue
        }
    }

    // Prim's Algorithm
    public static int primsMST(Map<Integer, List<Edge>> graph, int start) {
        boolean[] visited = new boolean[graph.size()]; // Track visited vertices
        PriorityQueue<Edge> pq = new PriorityQueue<>(); // Min-heap for smallest edges
        pq.add(new Edge(start, 0)); // Start from 'start' vertex with weight 0
        int totalCost = 0;

        while (!pq.isEmpty()) {
            Edge current = pq.poll(); // Get edge with smallest weight
            int v = current.vertex;
            int w = current.weight;

            if (visited[v]) continue; // Skip if vertex already visited

            visited[v] = true; // Mark vertex as visited
            totalCost += w; // Add edge weight to MST total cost

            // Explore all neighbors
            for (Edge neighbor : graph.get(v)) {
                if (!visited[neighbor.vertex]) {
                    pq.add(new Edge(neighbor.vertex, neighbor.weight)); // Add unvisited neighbor to heap
                }
            }
        }

        return totalCost; // Return total weight of MST
    }

    public static void main(String[] args) {
        // Graph represented as adjacency list
        Map<Integer, List<Edge>> graph = new HashMap<>();
        for (int i = 0; i < 4; i++) graph.put(i, new ArrayList<>());

        // Add edges (undirected graph)
        graph.get(0).add(new Edge(1, 2)); graph.get(1).add(new Edge(0, 2));
        graph.get(0).add(new Edge(2, 3)); graph.get(2).add(new Edge(0, 3));
        graph.get(0).add(new Edge(3, 1)); graph.get(3).add(new Edge(0, 1));
        graph.get(1).add(new Edge(3, 2)); graph.get(3).add(new Edge(1, 2));
        graph.get(2).add(new Edge(3, 4)); graph.get(3).add(new Edge(2, 4));

        int mstCost = primsMST(graph, 0); // Start from vertex 0
        System.out.println("Minimum Spanning Tree Cost: " + mstCost);
    }
}
